doc/technotes/errorHandling.rst

```diff
@@ -0,0 +1,190 @@
+.. _readme-errorHandling:
+
+==============
+Error Handling
+==============
+
+This README describes the error handling feature as currently implemented.
+
+
+Errors
+------
+A base class ``Error`` is defined in the Chapel standard modules.
+
+::
+
+  class Error {
+    var msg: string;
+
+    proc Error(_msg: string) {
+      msg = _msg;
+    }
+  }
+
+Users may use the provided ``Error`` directly, or create their own error
+hierarchy with ``Error`` at its root.
+
+::
+
+  class MyError: Error {}
+
+  class MyIntError: MyError {
+    var i: int;
+  }
+
+Chapel does not provide a standard set of ``Error`` classes.
+
+
+Throwing Errors
+---------------
+``throw`` statements throw errors from a function to its callee.
+
+::
+
+  throw new Error("with a message");
+
+  var e: Error;
+  // ... set e ...
+  throw e;
+
+If a function may throw an error, it must be declared with ``throws``.
+
+::
+
+  proc mayThrowErrors() throws { ... }
+  proc mayThrowErrorsAlso(): A throws where ... { ... }
+  proc mayNotThrowErrors() { ... }
+
+
+Handling Errors
+---------------
+
+Default Policy
+++++++++++++++
+
+When a throwing function is called, the default error handling policy of the
+calling function is as follows:
+
+* If the calling function is declared with ``throws``, any error encountered
+  will be thrown in turn.
+
+* If the calling function is not declared with ``throws``, the program will
+  ``halt()`` if any error is encountered.
+
+
+Specified Policy
+++++++++++++++++
+
+``try``/``try!`` blocks and their associated ``catch`` clauses allow the user
+to specify an explicit error handling policy for a throwing call.
+
+* Statements that contain throwing calls should be enclosed in a ``try`` block.
+  
+  * If an error is raised by a call in a ``try`` block, the rest of the block
+    is abandoned and control flow is passed to the list of ``catch`` clauses.
+ 
+* Each ``catch`` clause may contain an associated type filter. If the error's
+  type matches the type filter, the block associated with that ``catch`` clause
+  (and only that block) will be executed. 
+
+  * These filters will be evaluated in order of declaration, so even if a more
+    exact filter is present in the list, it will not be selected if a matching
+    general filter precedes it.
+
+  * If no filter is included, or only an error variable name is provided, then
+    all errors will match. Combined with the rule above, a clause with such a
+    filter must be declared at the end of the list.
```

Comment:
Combined with the rule above, a clause with such a filter must be declared at the end of the list. 

->

A clause with such a filter must be declared at the end of the list., since otherwise clauses that follow it could never be executed.